convert2rhel/redhatrelease.py

```diff
@@ -58,6 +56,26 @@ def get_system_release_content():
         loggerinst.critical("%s\n%s file is essential for running this tool." % (err, filepath))
 
 
+def get_os_release_filepath():
+    """Return path of the OS Release file."""
+    release_filepath = "/etc/os-release"  # RHEL 6/7/8 based OSes
+    if os.path.isfile(release_filepath):
+        return release_filepath
+    loggerinst.critical("Error: Unable to find the %s file containing the information needed." % release_filepath)
+
+
+def get_os_release_content():
+    """Return content of the file containing useful information
+    about the os release.
+    """
+
+    filepath = get_os_release_filepath()
+    try:
+        return utils.get_file_content(filepath)
+    except EnvironmentError as err:
+        loggerinst.critical("%s\n%s file is essential for running this tool." % (err, filepath))
+
```

Comment:
A few things I noticed about this block of code:

* `get_os_release_content()` doesn't appear to be used by anything?  (I thought `RestorableFile` might use it in some manner but I'm not seeing anything.)  Do we want to implement it if it's not being used?
  * Is os-release's info being used or are we just implementing these functions so that we can make it a `RestorableFile`?  If the latter, maybe we should use  `loggerinst.info()` (to match what RestorableFile.backup() uses) instead of loggerinst.critical()?
* The code to handle os-release seems to be largely the same as for system-release with just the hardcoded path changed.  It feels like we should parametrize that so we don't have to create new functions (maybe using functools.partial to keep the current API?)

A bit separate from your PR:
* it feels like exceptions should be raised by these functions (and RestorableFile).  The functions don't know whether the program should continue or exit if the file does not exist (or whether to log a missing file as critical or merely an unexpected but probably unimportant bit of info) so raising exceptions allows the caller to make that decision.  Maybe that's a change we should make in the future.
* We're checking that these filepaths exist quite a few times (eveytime get_os_release_filepath() is called explicitly; whenever get_os_release_content() is called, get_os_release_filepath() is called implicitly; and whenever RestorableFile.backup() is called the filepath is checked there as well.  At some point we probably want to change this code so that the file path is checked once near startup and then rely on raising an exception if the file doesn't exist later on.  (no matter what, there is always a race between when the file is checked and we actually open the file for use.  So checking near startup allows us to exit early in common cases.  Raising and catching exceptions allow us to handle the error when the file is used.)

Reply:
@abadger we have modified logger so critical level exits the program through a `sys.exit`, so there is an exception being thrown in `get_os_release_filepath`

Reply:
tldr;
* Exceptions say that the function doesn't know what to do so it is asking its callers to make the decision
* Exceptions need to communicate enough information about errors so that the caller can make that decision
* SystemExit is special because it is about control flow, not errors.

True, but catching SystemExit is probably not a generally good thing.  (I see that we do it in `main()` and that doesn't seem too bad to me but let me try to explain what the difference is).

At a high level, exceptions are an alternate method for the function to communicate with its caller from the normal return value.  They should be raised when a function is saying, "Hey, I encountered this error condition and I don't know how to proceed!".  Then each caller is able to catch the exception if they know how to proceed in case of the error.  If they don't know how to continue, they let the exception propagate further up the call stack.  If nothing handles the exception, then the user is notified and program execution stops.

Most of the time, a caller needs some bits of information in order to decide whether (and how) to handle the error case.  For instance, MemoryError (raised when the system can't allocate any more RAM to the program) might be a case where the caller cannot do anything because it's likely they won't be able to allocate any new data structures to aid in recovery.  OTOH, an IOError due to file system permissions might lead the caller to print a message to the user to change permissions and exit or to ignore the error because the file operation was an optional feature and doesn't interfere with the normal program flow.

In order to make a decision about whether an exception can be handled or not, the caller needs to know certain information about the error.  Some things the caller knows from its own context (for instance, that it was trying to backup an optional file).  Other things, the caller needs to receive from the function (for instance, that the error was because the file did not exist vs a permissions error vs memoryerror vs user aborting the process [etc]).  With all of that information, the caller can decide whether to continue or not.

Raising SystemExit throughout the code causes problems for this.  SystemExit is implemented as an exception but it's really a control flow structure (which is why the standard way to raise SystemExit is to call `sys.exit()` rather than `raise SystemExit()`).  It's message to the caller is "I've unilaterally decided that we need to exit so please unwind the stack and end now."  It's not really meant to be caught by callers.  For that reason, it doesn't give much information to the caller that would be useful to decide whether the caller should catch and process it or not. 

Tangent: We catch SystemExit in main().  Is that a bad thing?  I would say sort of but it's allowable.  We catch SystemExit as a way to inject an exit handler.  We want to make sure certain things happen when the program exits prematurely no matter what the reason for the exit is.  So we try to catch any and all reasons that the system could exit prematurely and run those handlers based on context that we know about in main() (at what point in the convert2rhel run we were at).  Using atexit() to do this might fit into the language's design better ( https://docs.python.org/3/library/atexit.html ) but I think this limited use of SystemExit isn't horrible.  Python makes it the most obvious way to create an exit handler so it doesn't feel too abusive to do it the way we are.

Back to the main topic... How does this apply to get_os_release_filepath(), RestorableFile, etc?  These functions and objects are checking whether the filepath that they are given exists and if so, they (indirectly) call sys.exit().  So they're saying "Hello caller, I've decided that the program needs to exit now.  Please do so".  When these were used only with mandatory files (needed  for the operation of convert2rhel), that choice didn't raise any red flags.  If you had raised a normal exception, the caller would have made the same decision that we're currently making in the callee.

With this PR, though, we're introducing optional RestorableFiles.  We want to make sure the file exists in its original state if we have to rollback but if they don't exist to begin with we're fine with that too.  That reveals the flaw with raising SystemExit in these functions.  We no longer want to exit in all cases.  Further, RestorableFile doesn't know which cases are mandatory and which are optional.  So we need the caller to make the decision on whether to exit or not.  What information does the function need to return to the caller?  The caller knows whether the file is optional or mandatory from its own context but it needs us to tell it what went wrong (FileNotFound on an optional file can be ignored.  PermissionError might point at some large, unexpected problem like chattr or a filesystem mounted read-only.  MemoryError is likely unrecoverable).  So instead of raising SystemExit, it's better if these functions raise an exception which contains the information the caller needs to make its decision to recover and continue or abort.

Reply:
Thanks for the insights. I wasn't aware of `atexit` and it looks like a better alternative to our exit mechanism indeed.

You raise good questions. We don't read the content of the /etc/os-release file. We just want to make sure we back it up if it exists. If it doesn't, it's ok, we shouldn't have a problem with that. So I agree with you on using `loggerinst.info()` instead of `loggerinst.critical()`. Perhaps with a message like _"Skipping backing up /etc/os-release - file not available."_.

Reply:
I'm getting on CentOS Linux 6:
```
Backing up None.
[12/03/2021 20:20:16] DEBUG - Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/convert2rhel/main.py", line 94, in main
    redhatrelease.os_release_file.backup()
  File "/usr/lib/python2.6/site-packages/convert2rhel/utils.py", line 527, in backup
    if os.path.isfile(self.filepath):
  File "/usr/lib64/python2.6/genericpath.py", line 29, in isfile
    st = os.stat(path)
TypeError: coercing to Unicode: need string or buffer, NoneType found

No changes were made to the system.
```

Reply:
The

Reply:
Fixed by https://github.com/oamg/convert2rhel/pull/384.